### ISSUE TITLE
Emoji Search

### ISSUE THREAD

**Issue by D3SOX** [florisboard/florisboard#45]
A search for emojis would be nice

**Comment by patrickgold** [comment]
Thanks for your suggestion. I guess you propose a search function similar to the one of GBoard, right? I'll add it to the todo list, probably for v0.4.0.

**Comment by HeiWiper** [comment]
I would like to suggest an addition which I'm not sure it is directly related to this. It would be nice to have emojis suggested based on input but only if the user starts the word with a specific character, for example the user would type `:<string>` and the suggestiom bar would display emojis which their labels match `<string>`.
This feature is similar to the one available on Discord and Anysoftkeyboard.

**Comment by D3SOX** [comment]
@HeiWiper Good idea! I think Gboard also has this when you type the exact emoji name (e.g. cat) and it suggests 🐈
Definitely related but I also think it doesn't exactly fit in here. We could open another issue for this

**Comment by patrickgold** [comment]
@HeiWipe I like your idea a lot, but to keep things tidy, it would be best if you made a separate issue for this, as this issue is for an emoji search box within the media UI. Thanks!

**Comment by Sitavi** [comment]
I'm agree with that, would be really great to have this feature. That's why anysoftkeyboard have an advantage for me

**Comment by JoeriTheGreat** [comment]
> I'm agree with that, would be really great to have this feature. That's why anysoftkeyboard have an advantage for me

I just tried anysoftkeyboard, but i think in terms of emoji suggestion i still like swiftkeyboard the best.
You can type anyting lik jo, be, ha -> normaly would be joy, beer, haha,... and you get a suggestion for a emoji.

**Comment by ZerataX** [comment]
I would recommend to not do what discord and slack does and use some weird custom list for emoji short codes and names.

while unicode codepoint names suck, there is unicode cldr which gives keywords and names in many languages for every emoji. it's a great resource and not as confusing as a lot of the names currently used in discord (e.g. blush and flushed are imo the opposite).

http://cldr.unicode.org/translation/characters-emoji-symbols/short-names-and-keywords

**Comment by sabzo** [comment]
I know this feature will be implemented on 0.5.0 -- however if one wanted to experiment and add it what are practical steps?

I was able to add an `EditText` searchbox into `smartbar.xml` and edit`smarbarview` to show the searchbox. However when I attempt to type into the smarbarview searchbox the text is added to an input connection that is outside of the keyboard. I thought maybe the answer might be to check `EditorInstance` and try to force the input connection to focus on the smartbar view search box but I couldn't find a way to do so. 

It's almost showing a keybaord within a keyboard -- or forcing an EditText in the MainArea of smartbarview to receive the textinput. 

Any thoughts? @patrickgold @Glitchy-Tozier

**Comment by patrickgold** [comment]
> I know this feature will be implemented on 0.5.0 -- however if one wanted to experiment and add it what are practical steps?

Well it is indeed tricky to have an edit text within the IME that is responsible for it. I'd suggest that the communication between the emoji edit text and the keyboard runs on an internal simulated input connection rather than trying to force the Android framework to focus on the edit text itself. 2 reasons for that:
1. If we force the focus away from the target app then we won't be able to re-focus it after the search without manual intervention by the user (I do not have the ability to give away focus from the keyboard, except when hiding the keyboard).
2. While the keyboard is showing for the internal edit text, a row of emojis is probably shown. Tapping on the suggested emojis should enter them into the target app rather than the edit text, so taking away the focus from the target app forcefully prevents us from doing that.

This means we have to kinda extend the EditText widget and allow for a "fake focus": When the user taps on the emoji search bar, the emoji edit text is visually prepped to look like it has focus and a cursor is also additionally rendered. When any character keystrokes or arrows, CTRL+V etc. is done, the emoji edit text receives this internally somehow - we could use a reference via the MediaManager and edit the characters manually for this or use Kotlin Co-routine Channels, or even implement a custom InputConnection - fact is, there are many ways to accomplish that. When an emoji is pressed however, it is processed normally and sent to the target app.

**Comment by Glitchy-Tozier** [comment]
> Any thoughts? @patrickgold @Glitchy-Tozier


Patrick is your guy for everything difficult, I'm not familiar enough with the code of the app

**Comment by sabzo** [comment]
> > I know this feature will be implemented on 0.5.0 -- however if one wanted to experiment and add it what are practical steps?
> 
> Well it is indeed tricky to have an edit text within the IME that is responsible for it. I'd suggest that the communication between the emoji edit text and the keyboard runs on an internal simulated input connection rather than trying to force the Android framework to focus on the edit text itself. 2 reasons for that:
> 
>     1. If we force the focus away from the target app then we won't be able to re-focus it after the search without manual intervention by the user (I do not have the ability to give away focus from the keyboard, except when hiding the keyboard).
> 
>     2. While the keyboard is showing for the internal edit text, a row of emojis is probably shown. Tapping on the suggested emojis should enter them into the target app rather than the edit text, so taking away the focus from the target app forcefully prevents us from doing that.
> 
> 
> This means we have to kinda extend the EditText widget and allow for a "fake focus": When the user taps on the emoji search bar, the emoji edit text is visually prepped to look like it has focus and a cursor is also additionally rendered. When any character keystrokes or arrows, CTRL+V etc. is done, the emoji edit text receives this internally somehow - we could use a reference via the MediaManager and edit the characters manually for this or use Kotlin Co-routine Channels, or even implement a custom InputConnection - fact is, there are many ways to accomplish that. When an emoji is pressed however, it is processed normally and sent to the target app.
> 
> This means we have to kinda extend the EditText widget and allow for a "fake focus": When the user taps on the emoji search bar, the emoji edit text is visually prepped to look like it has focus and a cursor is also additionally rendered. When any character keystrokes or arrows, CTRL+V etc. is done, the emoji edit text receives this internally somehow - we could use a reference via the MediaManager and edit the characters manually for this or use Kotlin Co-routine Channels, or even implement a custom InputConnection - fact is, there are many ways to accomplish that. When an emoji is pressed however, it is processed normally and sent to the target app.

Essentially it's mocking "edit text" functionality -- however it seems that you plan to redo the whole media component. So maybe I should wait for that phase? Or is the mocking code pretty independent from a future implementation?

**Comment by patrickgold** [comment]
> Essentially it's mocking "edit text" functionality -- however it seems that you plan to redo the whole media component. So maybe I should wait for that phase? Or is the mocking code pretty independent from a future implementation?

The mocking itself is pretty independent, because the internal way of consuming key events and passing around things won't change that much in 0.5.0. The biggest conflict point is anything which has directly to do with the UI (layout xml or custom widget code), but I think if you make the EditText as independent from the current emoji UI as possible then it will be easy to integrate your UI code in 0.5.0's rework.

**Comment by kik-btaski** [comment]
> > Essentially it's mocking "edit text" functionality -- however it seems that you plan to redo the whole media component. So maybe I should wait for that phase? Or is the mocking code pretty independent from a future implementation?
> 
> The mocking itself is pretty independent, because the internal way of consuming key events and passing around things won't change that much in 0.5.0. The biggest conflict point is anything which has directly to do with the UI (layout xml or custom widget code), but I think if you make the EditText as independent from the current emoji UI as possible then it will be easy to integrate your UI code in 0.5.0's rework.

What i understood is the problem of having 2 cursor parallely running on 2 edit texts. Am i right?

I think the other way would be to read the last word typed by user on any edit text and suggest emoji based on it when tapped on emoji tab. 

It would be like telegram suggesting stickers based on what last word typed🤔

**Comment by D3SOX** [comment]
Now that there are emoji name suggestions, this shouldn't be such a big effort

**Comment by patrickgold** [comment]
Thanks to the progress in implementing k3lp into FlorisBoard I have been able to experiment again with this feature and have implemented a technical proof-of-concept:

https://github.com/user-attachments/assets/40ba7603-c83e-409f-af45-6c8016d347f4

Now that the hard part (typing in a text field within the IME UI) is solved, it shouldn't take too long until this feature is properly implemented and released :)
